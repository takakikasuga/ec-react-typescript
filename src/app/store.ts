import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import counterReducer from '../features/counter/counterSlice';
import userReducer from '../features/user/userSlice';
import itemsReducer from '../features/items/itemsSlice';
import itemsCountReducer from '../features/itemCount/itemCountSlice';
import itemPriceReducer from '../features/itemPrice/itemPriceSlice';
import addOrderReducer from '../features/order/orderSlice';
import orderUpdateReducer from '../features/order/orderUpdateSlice';
import fetchOrderReducer from '../features/order/fetchOrderSlice';
import cartListsReducer from '../features/cartLists/cartListsSlice';
import statusZeroIdReducer from '../features/statusZeroId/statusZeroIdSlice';
import updateOrderStatusReducer from '../features/order/updateOrderStatusSlice';
import orderHistoryReducer from '../features/order/orderHistorySlice';
import suggestItemsReducer from '../features/suggest/suggestSlice';
import adminReducer from '../features/admin/adminSlice';
import adminItemsReducer from '../features/items/adminItemsSlice';
import localCartStrageReducer from '../features/cartLists/localCartStrageSlice';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    user: userReducer,
    items: itemsReducer,
    itemCount: itemsCountReducer,
    itemPrice: itemPriceReducer,
    addOrder: addOrderReducer,
    orderUpdate: orderUpdateReducer,
    fetchOrder: fetchOrderReducer,
    cartLists: cartListsReducer,
    statusZeroId: statusZeroIdReducer,
    updateOrderStatus: updateOrderStatusReducer,
    orderHistory: orderHistoryReducer,
    suggestItems: suggestItemsReducer,
    admin: adminReducer,
    adminItems: adminItemsReducer,
    localCartStrage: localCartStrageReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
