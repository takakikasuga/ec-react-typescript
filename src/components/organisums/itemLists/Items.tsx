import React from 'react';
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'

// スライサーより取得
import { selectItems, fetchItemsAsync } from '../../../features/items/itemsSlice'

// コンポーネント
import { ItemdetailPrice } from '../../molecules/itemPrice/ItemdetailPrice'
import { FavoriteIconHeart } from '../../atoms/icons/FavoriteIconHeart'
import { StarIcon } from '../../atoms/icons/StarIcon'
import { PrimaryButton } from '../../atoms/button/PrimaryButton'

// 型のインポート
import { fetchItems } from '../../../types/items/items'

// マテリアルUI
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';


const useStyles = makeStyles({
  root: {
    width: "30%",
    margin: "20px"
  },
  media: {
    height: 300,
  },
});
export const Items = () => {
  const items: fetchItems[] = useSelector(selectItems)
  const dispatch = useDispatch()
  const history = useHistory()


  const changeToDetail = (path: string) => {
    history.push(path)
  }

  // アイテム一覧を再表示する
  const displayItems = () => {
    dispatch(fetchItemsAsync())
  }

  const classes = useStyles();
  return (
    <>
      {!items.length ?
        <>
          <h1>該当する商品がございません。</h1>
          <div onClick={displayItems}>
            <PrimaryButton>商品一覧へ</PrimaryButton>
          </div>
        </>
        :
        <>
          <FlecItems>
            {items.map((item: any, index: number) => (
              <Card className={classes.root} key={index}>
                <CardActionArea key={index}>
                  <CardMedia
                    className={classes.media}
                    image={item.imagePath}
                    title={item.name}
                  />
                  <CardContent>
                    <Typography gutterBottom variant="h5" component="h2">
                      {item.name}
                    </Typography>
                    <ItemdetailPrice item={item}></ItemdetailPrice>
                  </CardContent>
                </CardActionArea>
                <CardActions>
                  <FavoriteIconHeart></FavoriteIconHeart>
                  <StarIcon></StarIcon>
                </CardActions>
                <div onClick={() => { changeToDetail(`/detail/${item.id}`) }}>
                  <PrimaryButton>詳細へ</PrimaryButton>
                </div>
              </Card>
            ))}
          </FlecItems>
          <div onClick={displayItems}>
            <PrimaryButton>商品一覧へ</PrimaryButton>
          </div>
        </>
      }
    </>
  );

}

// スタイルコンポーネント
const FlecItems = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: flex-start;
`

