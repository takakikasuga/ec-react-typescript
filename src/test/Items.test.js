import React from "react"
import { render, cleanup, screen, fireEvent, waitFor } from "@testing-library/react"
import { afterEach, expect } from "@jest/globals"
import userEvent from "@testing-library/user-event"
import { Items } from "../components/organisums/itemLists/Items"
import { store } from "../app/store"
import { Provider } from "react-redux"


describe("Itemsコンポーネント", () => {

  // 各処理が通った後にアンマウントさせる
  afterEach(() => cleanup())

  afterEach(cleanup)

  // 各テストの直前に実行される（コンポーネントの取得）
  // beforeEach(() => {
  //   render(
  //     <Provider store={store}>
  //       <Items />
  //     </Provider>
  //   )
  // })

  describe("特定の文字列を含んでいるか否かの確認", () => {
    test("グローバルのアイテム情報を取得した場合に表示される「詳細」という文字列が初回レンダリング時には存在しないこと", () => {
      // この要素がないことを確かめる
      expect(screen.queryByText(/詳細/)).toBeNull()
    })
    test("グローバルのアイテム情報を取得した場合に表示される「商品一覧へ」という文字列が初回レンダリング時には存在すること", () => {
      // スクリーンオブジェクトでテキストを取得する
      screen.getByText("商品一覧へ")
      expect(screen.getByText("商品一覧へ")).toBeInTheDocument()
    })
    test("いかなる状況下でも「Searches for JavaScript」という文字列が存在しないこと", () => {
      // スクリーンオブジェクトでテキストを取得する
      // HTMLタグで取得して、その中身の存在を確認する
      expect(screen.queryByText(/Searches for JavaScript/)).toBeNull();
    })
  })
  // .only修飾子で１つのみ実行する
  describe("指定したHTMLタグが存在するか否かの確認", () => {
    // test("buttonタグを取得して、type属性が含まれているかの確認", () => {
    //   expect(screen.getByRole("button")).toHaveAttribute("type")
    // })
    test("buttonタグを取得してその中に文字列が含まれているかの確認", () => {
      // スクリーンオブジェクトでテキストを取得する
      // HTMLタグで取得して、その中身の存在wを確認する
      expect(screen.getByRole("button")).toBeInTheDocument()
      screen.debug()
    })
    test("h1タグを取得してその中に文字列が含まれているかの確認", () => {
      // スクリーンオブジェクトでテキストを取得する
      // HTMLタグで取得して、その中身の存在を確認する
      expect(screen.getByRole("heading")).toBeInTheDocument()
      // screen.debug()
    })
  })
  describe("指定したHTMLタグが要素数を確認", () => {
    test("buttonタグの数を確認（特定数）", () => {
      expect(screen.getAllByRole("button")).toHaveLength(1);
    })
  })
  describe("レンダリングされているかを確認", () => {
    test("全ての要素が正しくレンダリングされているのか否か", () => {
      screen.debug()
      expect(screen.getAllByRole("button")).toBeTruthy();
    })
  })
  describe("レンダリングされているかを確認", () => {
    test(" 関数が発火しているか否か", () => {

      const outputConsole = jest.fn()
      // 関数が発火しているか否か
      render(
        <Provider store={store}>
          <Items outputConsole={outputConsole} />
        </Provider>
      )
      // screen.debug()
      userEvent.click(screen.getByRole("button"))
      expect(outputConsole).toHaveBeenCalled()
    })
  })
  describe.only("レンダリングされているかを確認", () => {
    test("data-testidを使った発火テスト", () => {

      const outputConsole = jest.fn()
      // 関数が発火しているか否か
      render(
        <Provider store={store}>
          <Items outputConsole={outputConsole} />
        </Provider>
      )
      screen.debug()
      userEvent.click(screen.getByTestId("発火テスト"))
      expect(outputConsole).toHaveBeenCalled()
    })
  })
})