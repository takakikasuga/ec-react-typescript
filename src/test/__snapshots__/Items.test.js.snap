// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <h1>
    該当する商品がございません。
  </h1>,
  <div
    data-testid="発火テスト"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        商品一覧へ
      </span>
    </button>
  </div>,
]
`;
